File: repositories/qualification_repository.go
package repositories

import (
	"context"

	"doctor-service/entities"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type QualificationRepository interface {
	Create(ctx context.Context, qual *entities.DoctorQualification) error
	FindByDoctorID(ctx context.Context, doctorID uuid.UUID) ([]entities.DoctorQualification, error)
	Delete(ctx context.Context, qualificationID, doctorID uuid.UUID) error
}

type GormQualificationRepository struct {
	db *gorm.DB
}

func NewQualificationRepository(db *gorm.DB) QualificationRepository {
	return &GormQualificationRepository{db: db}
}

func (r *GormQualificationRepository) Create(ctx context.Context, qual *entities.DoctorQualification) error {
	return r.db.WithContext(ctx).Create(qual).Error
}

func (r *GormQualificationRepository) FindByDoctorID(ctx context.Context, doctorID uuid.UUID) ([]entities.DoctorQualification, error) {
	var quals []entities.DoctorQualification
	err := r.db.WithContext(ctx).Where("doctor_id = ?", doctorID).Find(&quals).Error
	return quals, err
}

func (r *GormQualificationRepository) Delete(ctx context.Context, qualificationID, doctorID uuid.UUID) error {
	return r.db.WithContext(ctx).Where("qualification_id = ? AND doctor_id = ?", qualificationID, doctorID).Delete(&entities.DoctorQualification{}).Error
}

File: python_backend/repositories/qualification_repository.py
from typing import List
from uuid import UUID
from sqlalchemy.orm import Session
from entities.qualification import DoctorQualification

class QualificationRepository:
    def __init__(self, db: Session):
        self.db = db

    def Create(self, qual: DoctorQualification) -> None:
        self.db.add(qual)
        self.db.commit()
        self.db.refresh(qual)

    def FindByDoctorID(self, doctor_id: UUID) -> List[DoctorQualification]:
        return self.db.query(DoctorQualification).filter(
            DoctorQualification.doctor_id == str(doctor_id)
        ).all()

    def Delete(self, qualification_id: UUID, doctor_id: UUID) -> None:
        self.db.query(DoctorQualification).filter(
            DoctorQualification.qualification_id == str(qualification_id),
            DoctorQualification.doctor_id == str(doctor_id),
        ).delete()
        self.db.commit()
